content/pixel_fifo.md

```diff
@@ -0,0 +1,252 @@
+# Pixel FIFO
+Before we get started, all references to a cycle are meant as T-cycles
+(4.19 MHz) and cycle counts are doubled on CGB in double speed mode. Also
+when stating that a certain action "lengthens mode 3" this means that
+mode 0 (hblank) is shortened. The extra time has to be made up somewhere
+and the lengthening of mode 3 eats into the mode that comes afterwards,
+which happens to be mode 0 as shown in the following diagram.
+
+![](imgs/game-boy-lcd-refresh-diagram.png "imgs/game-boy-lcd-refresh-diagram.png")
+
+## FIFO info
+FIFO stands for First In, First Out. The first pixel to be pushed to the
+FIFO is the first pixel to be popped off. In theory that sounds great,
+in practice there are a lot of intricacies.
+
+There are two pixel FIFOs. One for background pixels and one for OAM
+(sprite) pixels. These two FIFOs are not shared. They are independent
+of each other. The two FIFOs are mixed only when popping items. Sprites
+take priority unless they're transparent (color 0) which will be
+explained in detail later. The FIFOs can each hold up to 16 pixels at a
+time. The two FIFOs are manipulated only during mode 3 (pixel transfer)
+and they are both cleared at the beginning of mode 3.
```

Comment:
This is confusing. If they are cleared at the start of mode 3 that implies they are empty so Pixel Rendering cannot take place. Surely its cleared and the fetcher starts 8/16 cycles before model 3 so that there are pixels ready at the start of model 3?

Reply:
I think this might be implementation specific to SameBoy. There is code for clearing both FIFOs, filling them with "junk" and resetting the fetcher at the beginning of mode 3. It might be safe to assume this does not represent the behavior of the real hardware. Reading the code again, I think this might even only be executed at the beginning of each frame.